server/routes/queries.js

```diff
@@ -866,6 +865,72 @@ router.get("/eeris", canAccessForCollection, (req, res) => Project.findOne({ pro
+* @apiVersion 0.1.0
+* @apiName eerisHistAvgCustom
+* @apiGroup Queries
+* @apiParam {String} PROJECT_ID Project's unique ID.
+* @apiParam {String} readKey/masterKey Key for authorized read.
+* @apiParam {String} event_collection Event collection.<br/><strong><u>Note:</u></strong> Event collection names must start with a
+* letter and can contain only lowercase letters and numbers.
+* @apiParam {String} target_property Desired Event collection's property.<br/><strong><u>Note:</u></strong> Property names must start with a
+* letter and can contain only lowercase letters and numbers.
+* @apiParam {String} installationId ID of the installation
+* @apiParam {String} startDate Start date of the interval (YYY-MM-DD)
+* @apiParam {String} endDate End date of the interval (YYYY-MM-DD)
+* @apiSuccess {Boolean} ok If the query succeded.
+* @apiSuccess {Array} results Query result.
+* @apiSuccessExample {json} Success-Response:
+*     HTTP/1.1 200 SUCCESS
+*     {
+*       "ok": true
+*       "results": {
+*        "values": [...],
+*       }
+*     }
+* @apiUse NoCredentialsSentError
+* @apiUse KeyNotAuthorizedError
+* @apiUse ProjectNotFoundError
+* @apiUse TargetNotProvidedError
+* @apiUse BadQueryError
+*/
+router.get("/eerisCustom", canAccessForCollection, (req, res) => Project.findOne({ projectId: req.params.PROJECT_ID }).lean()
+  .exec(async (err2, project) => {
+    try {
+      if (err2 || !project) return res.status(404).json({ ok: false, results: "ProjectNotFoundError" });
+      const { readKey, masterKey, event_collection, target_property, startDate, endDate } = req.query;
+      if (!target_property) return res.status(400).json({ ok: false, results: "TargetNotProvidedError" });
+      if (!(readKey === project.readKey || masterKey === project.masterKey)) {
+        return res.status(401).json({ ok: false, results: "KeyNotAuthorizedError" });
+      }
+      // Check validity of date parameters
+      if (!startDate || !endDate) return res.status(400).json({ ok: false, results: "BadQueryError", message: "`startDate` and `endDate` are required" });
+      const date = new Date(startDate);
+      const finalDate = new Date(endDate);
+      if (date.getTime() > finalDate.getTime()) return res.status(400).json({ ok: false, results: "BadQueryError", message: "`startDate` must precede `endDate`" });
+
+      const keyName = `${req.params.PROJECT_ID}_${event_collection}_${target_property}_hist`;
+      const values = [];
+      const value = await r.get(keyName);
+      const jsonValue = JSON.parse(value);
+      finalDate.setDate(finalDate.getDate() + 1);
+      while (date.getTime() !== finalDate.getTime()) {
+        const year = date.getFullYear();
+        const month = (`0${date.getMonth() + 1}`).slice(-2);
+        const day = (`0${date.getDate()}`).slice(-2);
+        const avg = jsonValue[`avg_${year}-${month}-${day}`] || 0;
+        values.push(avg);
+        date.setDate(date.getDate() + 1);
+      }
```

Comment:
`getTime` returns the UNIX timestamp in milliseconds, but in each iteration you are only changing the day of the month. Won't this lead to an infinite loop?

Reply:
The change of day will propagate to a change in month and year accordingly (see screenshot below). However, the comparison operator "!==" should be replaced by "<", I think it's safer.

![date](https://user-images.githubusercontent.com/8168416/97931265-e842c080-1d75-11eb-8114-6fe4bbb09b02.png)

Reply:
It does, but `getTime()` also takes into account the time:

```js
new Date("2020-11-08T06:47:02.609Z").getTime() !== new Date("2020-11-08T05:00:01.406Z").getTime()
```

However, I just noticed that this endpoint accepts dates as `YYYY-MM_DD` and not as `ISOString`, like every other (in the `timeframe` parameter). In that case the time will default to `00:00:00.000Z`, so thats why this happens to work correctly.